### PR TITLE
Oracle GoldenGate on Docker

### PR DESCRIPTION
**Pull request by sbalousek**
- Support for Oracle GoldenGate version 21.3.0.0.0 for Oracle

Signed-off-by: Stephen Balousek <stephen.balousek@oracle.com>

### CODE REVIEW

File: OracleGoldenGate/21c/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,27 @@
+## Copyright (c) 2021, Oracle and/or its affiliates.
+FROM          oraclelinux:8-slim
```
Comment (Djelibeybi) [inline OracleGoldenGate/21c/Dockerfile:2]: We do not recommend the use of `oraclelinux:8-slim` for anything except microservices. Please rebase this to use `oraclelinux:8` instead.

File: OracleGoldenGate/21c/README.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+All shell commands in this document assume the usage of Bash shell.
+
+For more information about Oracle GoldenGate please see the [Oracle GoldenGate 21.3 On-line Documentation](https://docs.oracle.com/en/middleware/goldengate/core/21.3/index.html).
+
+## Build an Oracle GoldenGate Docker Image
+
+Once you have downloaded the Oracle GoldenGate software, a Docker image is created using the standard Docker command line.
+A single `--build-arg` is needed to indicate the GoldenGate installer which was downloaded.
+
+    $ docker build --tag=oracle/goldengate:21.3.0.0.0 \
```
Comment (Djelibeybi) [inline OracleGoldenGate/21c/README.md:37]: Please wrap code blocks with triple ticks and specify the syntax for better code highlighting, ie. 
```
` ` `shell
$ docker build --tag=oracle/goldengate:21.3.0.0.0 \
               --build-arg INSTALLER=fbo_ggs_Linux_x64_Oracle_services_shiphome.zip .
` ` `
```

Without the spaces, of course. :)

File: OracleGoldenGate/21c/install-prerequisites.sh [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+function success() {
+    echo "Packages installed after ${sequence} attempts"
+    rm -rf /var/cache/yum /var/log/yum* /tmp/yum*
+    useradd ogg
+    exit 0
+}
+
+for sequence in $(seq 3); do
+    "${manager}" -y install "${packages[@]}" && success "${sequence}"
```
Comment (Djelibeybi) [inline OracleGoldenGate/21c/install-prerequisites.sh:21]: You should be using modules for OL8 which is annoying with `microdnf` and why we don't recommend it. After rebasing to `oraclelinux:8`, please use `dnf module install nginx` and `dnf module install python36` for those components.

**Review by Djelibeybi — CHANGES_REQUESTED** [review]
It's my fault for not updating the contribution guide (which I'll do today) but we don't recommend using `oraclelinux:8-slim` for anything except statically built microservice binaries written in Go or GraalVM native images. Essentially anything that needs to install additional packages should use the `oraclelinux:8` and `dnf`.

File: OracleGoldenGate/21c/README.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+## Before You Start
+
+This project was tested with:
+
+* Oracle GoldenGate 21.3.0.0.0 Microservices for Oracle
+
+Support for Oracle GoldenGate Classic Architecture is not provided.
+
+**IMPORTANT:** You must download the installation binaries of Oracle GoldenGate. You only need to provide the binaries for the version you plan to install. The binaries can be downloaded from the [Oracle Technology Network](http://www.oracle.com/technetwork/middleware/goldengate/downloads/index.html). Do not decompress the Oracle GoldenGate ZIP file. The `dockerBuild.sh` script will handle that for you. You also must have Internet connectivity when building the Docker image for the package manager to perform additional software installations.
```
Comment (Djelibeybi) [inline OracleGoldenGate/21c/README.md:26]: Please change Docker to container. It would be great if `dockerBuild.sh` became `containerBuild.sh` too, particularly as OL8 now uses Podman.

File: OracleGoldenGate/21c/README.md [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+All shell commands in this document assume the usage of Bash shell.
+
+For more information about Oracle GoldenGate please see the [Oracle GoldenGate 21.3 On-line Documentation](https://docs.oracle.com/en/middleware/goldengate/core/21.3/index.html).
+
+## Build an Oracle GoldenGate Docker Image
+
+Once you have downloaded the Oracle GoldenGate software, a Docker image is created using the standard Docker command line.
```
Comment (Djelibeybi) [inline OracleGoldenGate/21c/README.md:34]: Change the first instance of Docker to container, i.e. "... a container image can be created using the Docker command-line interface.`

File: OracleGoldenGate/21c/README.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+```sh
+$ docker build --tag=oracle/goldengate:21.3.0.0.0 \
+                --build-arg INSTALLER=fbo_ggs_Linux_x64_Oracle_services_shiphome.zip .
+Sending build context to Docker daemon
+...
+Successfully tagged oracle/goldengate:21.3.0.0.0
+```
+
+## Running Oracle GoldenGate in a Docker container
```
Comment (Djelibeybi) [inline OracleGoldenGate/21c/README.md:45]: Remove the word Docker, please; "in a container" is sufficient.

File: OracleGoldenGate/21c/README.md [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+To run your Oracle GoldenGate Docker image use the `docker run` command as follows:
```
Comment (Djelibeybi) [inline OracleGoldenGate/21c/README.md:47]: Rephrase this as "Use the `docker run` command to create and start a container from the Oracle GoldenGate container image."

File: OracleGoldenGate/21c/README.md [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+Version 21.3.0.0.0 ...
+```
+
+## Known Issues
+
+None
+
+## License
+
+All scripts and files hosted in this project and GitHub [docker-images/OracleGoldenGate](../) repository required to build the Docker images are, unless otherwise noted, released under the Universal Permissive License (UPL), Version 1.0.  See [LICENSE](../LICENSE) for details.
```
Comment (Djelibeybi) [inline OracleGoldenGate/21c/README.md:145]: Replace Docker with container and link to the `LICENSE` file in the root of the repo.

File: OracleGoldenGate/21c/README.md [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+To download and run Oracle GoldenGate, regardless whether inside or outside a Docker container, you must download the binaries from the [Oracle Technology Network](http://www.oracle.com/technetwork/middleware/goldengate/downloads/index.html) and accept the license indicated at that page.
```
Comment (Djelibeybi) [inline OracleGoldenGate/21c/README.md:147]: Remove the word Docker, please.

File: OracleGoldenGate/21c/install-prerequisites.sh [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+packages=(java-1.8.0-openjdk jq libaio libnsl python3-requests python3-psutil tar unzip)
+
+function success() {
+    echo "Packages installed after ${sequence} attempts"
+    rm -rf /var/cache/yum /var/log/yum* /tmp/yum*
+    useradd ogg
+    exit 0
+}
+
+for sequence in $(seq 3); do
```
Comment (Djelibeybi) [inline OracleGoldenGate/21c/install-prerequisites.sh:19]: Not a merge blocker, but I'd like to know what you've experienced that's required you add code to try this command three times.

File: OracleGoldenGate/README.md [hunk 11]
```diff
@@ -1,8 +1,13 @@
-Oracle GoldenGate on Docker
-===============
+# Oracle GoldenGate on Docker
```
Comment (Djelibeybi) [inline OracleGoldenGate/README.md:1]: Please replace "on Docker" with "in a container"

File: OracleGoldenGate/README.md [hunk 12]
```diff
@@ -1,8 +1,13 @@
...
+
 Sample Docker build files to provide an installation of Oracle GoldenGate for DevOps users.
 
 ## Oracle GoldenGate 19c and earlier
 
 Provides Docker build files to create a Docker image for Oracle GoldenGate versions 12c through 19c.
```
Comment (Djelibeybi) [inline OracleGoldenGate/README.md:7]: Please replace Docker with container for the second instance, i.e. "Provides Docker build files to create a container image..."

File: OracleGoldenGate/README.md [hunk 13]
```diff
@@ -1,8 +1,13 @@
...
 For more details, see [12c-19c/README.md](12c-19c/README.md).
+
+## Oracle GoldenGate 21c
+
+Provides Docker build files to create a Docker image for Oracle GoldenGate version 21c.
```
Comment (Djelibeybi) [inline OracleGoldenGate/README.md:12]: Please replace Docker with container for the second instance, i.e. "Provides Docker build files to create a container image..."

**Review by Djelibeybi — CHANGES_REQUESTED** [review]

**Comment by sbalousek** [comment]
Reworked the README verbiage with respect to 'Docker'.  I've not tested with `podman`, though. Would you like to see the examples reworked to illustrate both container runtimes?

File: OracleGoldenGate/21c/README.md (hunk 8 above)
Comment (sbalousek) [reply OracleGoldenGate/21c/README.md]: `LICENSE` is already referenced here.  Am I missing something?

Comment (sbalousek) [reply OracleGoldenGate/21c/README.md]: I changed the references to `LICENSE` and removed that file from the `OracleGoldenGate` directory.

**Review by Djelibeybi — APPROVED** [review]
